### ISSUE TITLE
ML tags sum to 256, not 255

### ISSUE THREAD

**Issue by amurray411** [nanoporetech/dorado#1495]
### New issue checks

- [x] I have read the [Dorado Documentation](https://dorado-docs.readthedocs.io/en/latest/).
- [x] I did not find an [existing issue](https://github.com/nanoporetech/dorado/issues?q=is%3Aissue).

### Dorado version

```true
0.6.0 to 0.9.6
```

### Dorado subcommand

Basecaller

### The issue

I've been using dorado basecaller both through minknow aka dorado basecall server (7.4.12 through 7.9.8) and through post-sequencing basecalling, utilizing the basecalling models "dna_r10.4.1_e8.2_400bps_sup@v4.3.0" and 
"dna_r10.4.1_e8.2_400bps_sup@v5.0.0", the latter only being used with the recent 0.9.6 functionality implemented by 7.9.8.

In all cases, I've been generating methylation data for 5hmCG and 5mCG only in CG contexts. **The problem I'm having is that ML tags for certain CG sites on certain reads are reporting 5hmC and 5mCG integers that add up to 256 rather than 255.** My understanding is that these "probabilities" of modification are reported between 0 and 255.

However, I'm seeing two distinct scenarios that draw my attention: **1)** the 5hmC probability integer is 1, not 0, when the corresponding 5mC probability is 255, and **2)** the 5hmC probability integer is 129 when the corresponding 5mC probability is 127. I can definitely understand ML tags of 127 and 128, respectively, where an ML tag can't represent exact equality of likelihood, but I'm unsure where the 127 and 129 is originating from/how to adjust for ML tags totaling to 256. Are these scenarios supposed to be 127 and 128 instead? Also, if the sum of all ML tags for a certain read is supposed to be 256, why are there no cases in 100 million-ish CG calls where the probabilities of 5hmC and 5mC are 128 and 128? I suspect it has something to do with the floating point precision behind the dorado basecalling training, but I'm unsure of how to deal with that in my analyses.

The reason I'm noticing this problem is that I've recently begun analysis on h and m mods individually, rather than the typical modkit adjust-mods --convert h m where methylation tags are combined into a single probability. To do this analysis on h and m separately, I'm using modkit extract full and noticed some negative canonical C probabilities, driven by h + m being greater than 1. I initially thought this was a modkit issue, but after diving into the modkit code, I've realized that their errors have come from adjustment of the dorado-produced bam files. While I can adjust the ML tag integers manually for a simple global methylation analysis, I'm really interested in the raw truth (or any sort of explicit 5hmCG/5mCG validation) that is output by the dorado basecaller algorithm, especially since I read a discussion on 5hmCG calls produced as false positives for 5mCG calls, which was raised as an issue(question) on the modkit github.

I will also acknowledge that it's possible a later basecalling model addresses the problem I'm observing; however, in the context of a clinically-relevant timeline, I feel that it's important for me to use dorado basecall server for real-time basecalling of high depth data.

### System specifications

Linux
Nvidia Geforce 3090 multi-GPU programming
Slurm for post-dorado analysis using samtools and modkit

**Comment by malton-ont** [comment]
Hi @amurray411,

Dorado modbase calling generates a set of scores (one for each mod and one for the canonical base), and then this set is passed through a softmax layer, which effectively converts each score to a probability (summing to 1). In order to store these probabilities in the BAM format they are then binned into an integer value using the following code:

https://github.com/nanoporetech/dorado/blob/e72f14925cd435fff823ebf244ce2195b135a863/dorado/read_pipeline/nodes/ModBaseCallerNode.cpp#L558-L559

What you are seeing is probably due to floating point precision - the floating point values of the probabilities cannot be represented exactly, so we end up with a situation where `P(C) + P(m) + P(h) = 1.00001` or similar. Without digging into the intermediate values I can't be more precise about the exact conversions occurring.

**Comment by ArtRand** [comment]
Hello @amurray411, 

Does the BAM you attached to [this issue](https://github.com/nanoporetech/modkit/issues/474) expose the problem? Modkit should be able to handle this kind of thing.

**Comment by amurray411** [comment]
Hello @malton-ont and @ArtRand,

Thank you both for the quick responses. I've been looking through this: https://github.com/nanoporetech/dorado/blob/e72f14925cd435fff823ebf244ce2195b135a863/dorado/read_pipeline/nodes/ModBaseCallerNode.cpp#L558-L559 and also ModBaseChunkCallerNode.cpp this morning, but since I am not fluent in C++, it's been a bit of a challenge to identify the source of the problem.

> Does the BAM you attached to [this issue](https://github.com/nanoporetech/modkit/issues/474) expose the problem? Modkit should be able to handle this kind of thing.

Yes, the aforementioned BAM (1 read) has multiple offenders in that the ML h tag and ML m tag add up to 256 exactly, making the P(C) + P(h) + P(m) = 1.003906. In this read, there are cases in which modkit centers the bin incorrectly, such that where the ML tag for m is 255, modkit centers the bin to be 255.5, a value that I believe does not exist within the 8-bit range of 0-255. This wouldn't be a problem in most cases where the ML tag for one mod is 255 and for the other is 0, but I am seeing cases where the h ML tag is 1 and the m ML tag is 255, so centering the bins to 1.5 and 255.5 actually makes the total 257. I've fixed this specific issue by personally centering the bin from 255.5 (output form modkit extract full) to 254.5. 

However, in cases where the ML tags are 134 and 122 (or 127 and 129 such as in other reads not included in this BAM file), I am unsure which tag (h or m) is being incorrectly adjusted/rounded/coerced from the softmax layer in dorado. Thus, in order to center the relevant bins in the same way that modkit does, I am unsure which of the ML tags to assume is incorrectly binned to the higher integer. Again, in these cases, modkit extract full adjusts these values to 134.5 and 122.5, effectively summing to 257, but I feel that if I adjust both integers to the center of the lower bin (aka to 133.5 and 121.5), I am then introducing an integer of 1 as the inferred canonical C probability, which I know to be incorrect.

This is why I was attempting to use the middle cases to identify the offending ML tag; I was hoping to gain an understanding of which tag (h or m) was more likely to be incorrectly adjusted/rounded/coerced. But it seems, at least from my perspective, that it is random. If precision error is truly the problem (I suspect that it is), then that explains the randomness. However, I am still not sure how modkit should/could/would center the bins in these scenarios.

I am happy to add a read or two more with centered corner cases if that helps identify the cause. These small adjustments also may not even matter in the grand scheme of things, but as a graduate student, my curiosity is getting the best of me.

Thanks,
Allison

**Comment by marcus1487** [comment]
Thanks for the detailed write-up and examples. While floating point precision is likely the root cause here, I think the main issue is how binning affects the ML tag encoding in the BAM format.

The ML field stores probabilities using 8-bit bins:
* Bin 0 represents values between 0/256 and 1/256
* Bin 1 represents values between 1/256 and 2/256
… and so on up to:
* Bin 255, which represents values between 255/256 and 256/256

Because of this, modkit correctly centers bin 255 at 255.5. But the key point is that this encoding is lossy; there isn’t a one-to-one mapping back to the original float values.

That’s why combinations like (127, 129) appear. They’re valid under the binning scheme even though they can look “impossible” probabilistically, since the bins are only approximations of the true softmax outputs. The only case where we can perfectly recover the underlying values is when the bins sum to 256. In those instances, the correct interpretation is that 127 means exactly 127/256 and 129 means exactly 129/256, which sum cleanly to 1 (and the canonical base probability must have been very close to 0). Note that even sums to probability greater than 1 are valid for BAMS produced using models for more than 2 modified bases (again due to binning/rounding issues).

Why (128, 128) never occurs is harder to pin down. It may be related to how Dorado’s internals handle ties during softmax + binning, but that’s just speculation.

So the short version is: values summing to 256 are valid and come from the lossy nature of BAM storage. Modkit’s centering works consistently with this scheme, though you’ll sometimes see totals exceeding 1 due to the unavoidable rounding.

**Comment by amurray411** [comment]
Hello,

Sorry for the delay, but I still have three quick clarifying points:

1, Thank you for this description of the binning. I understand now how 255.5 can exist in the probability space even though it is outside of the integer space. 

I had followed the samtools definition of the ML tags (https://github.com/samtools/hts-specs/blob/master/SAMtags.pdf) such that the true probability is within the range of N/256 and (N+1)/256, but incorrectly assumed that referred to total modifications' probability, not each mod's probability. On the same note, thank you for the 127 + 129 example. I now understand enough to decide what to do in most of these cases, especially where the ML tags add up to 256. 

That being said, in these scenarios where the exact probability is certain, doesn't it make the most sense for modkit (or me) to keep 127 as 127 and 129 as 129, rather than centering them to 127.5 and 129.5 where the total is 257? or am I still misunderstanding? @ArtRand ?

> Note that even sums to probability greater than 1 are valid for BAMS produced using models for more than 2 modified bases (again due to binning/rounding issues).

This is a great point, thank you. I don't currently run any models with greater than 2 modified bases included, but I will certainly take note of this. 

> Why (128, 128) never occurs is harder to pin down. It may be related to how Dorado’s internals handle ties during softmax + binning, but that’s just speculation.

2, After further analysis, I was wrong. I had introduced my own floating point errors in my analysis and incorrectly filtered out these scenarios. Although rare (2 CG calls in 15 million CG calls), I have found instances that (128, 128) does occur, although again, modkit centers the values to 128.5 and 128.5. But I just wanted to follow up to say that exact ties do occur, I just skipped over them. 

> So the short version is: values summing to 256 are valid and come from the lossy nature of BAM storage. Modkit’s centering works consistently with this scheme, though you’ll sometimes see totals exceeding 1 due to the unavoidable rounding.

3, I was wondering if there's any way I could (if I really wanted to or needed to throughout this project) retrieve at least 16-bit values of probabilities before coercion into BAM format, or alternatively, generate a pseudoBAM format to store these values, at least for very small BAM files. This is purely hypothetical, but based on my future plans, I suspect I may need the increased precision. Can I make small changes to my local version of Dorado to do so or is that well beyond the scope of Dorado?

Thanks again for all of your insight. I really appreciate you all taking the time to discuss this with me, and if you could just follow up briefly on 1 and 3, I'm happy to close the inquiry.

Best,
Allison

**Comment by malton-ont** [comment]
> 3, I was wondering if there's any way I could (if I really wanted to or needed to throughout this project) retrieve at least 16-bit values of probabilities before coercion into BAM format, or alternatively, generate a pseudoBAM format to store these values, at least for very small BAM files. This is purely hypothetical, but based on my future plans, I suspect I may need the increased precision. Can I make small changes to my local version of Dorado to do so or is that well beyond the scope of Dorado?
> 

Hi @amurray411,

You're welcome to fork dorado and modify it for your needs. You'll want to add a new member to store the floating point probabilities [here](https://github.com/nanoporetech/dorado/blob/e72f14925cd435fff823ebf244ce2195b135a863/dorado/read_pipeline/base/include/read_pipeline/base/messages.h#L58), gather the chunk scores [here](https://github.com/nanoporetech/dorado/blob/release-v1.1/dorado/read_pipeline/nodes/ModBaseCallerNode.cpp#L558) and probably output them to a new tag [here](https://github.com/nanoporetech/dorado/blob/e72f14925cd435fff823ebf244ce2195b135a863/dorado/read_pipeline/base/messages.cpp#L307).

**Comment by ArtRand** [comment]
Hello @amurray411, 

I think it's a bug that Modkit output probabilities are not valid probabilities, but I need to come up with a sensible and internally consistent way to handle this. I think for the users of `modkit extract` functions is probably makes the most sense to renormalize the probabilities when they sum to >1, but I want to make sure doing such a transformation doesn't add to confusion.
